File: lib/py-edge-driver/src/amrc/factoryplus/edge_driver/handler.py
# Copyright (c) University of Sheffield AMRC 2025.

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING, ClassVar, Optional, Set, Any, Dict, Callable, Union, Awaitable

if TYPE_CHECKING:
    from .driver import Driver


class Handler(ABC):
    """
    Abstract base class defining the interface for device handlers.

    Any class implementing this interface can be used as a device handler
    with the Driver class. This class cannot be instantiated directly.
    """

    validAddrs: ClassVar[Optional[Set[Any]]] = None
    """Set of valid addresses that this handler can manage, or None if any address is valid."""

    def parseAddr(self, addr: Any) -> Any:
        """
        Parse a device address into a format usable by the handler.

        Args:
            addr: The raw address to parse

        Returns:
            The parsed address, or a falsy value if parsing failed
        """
        return addr

    @abstractmethod
    def close(self, callback: Optional[Callable[[], None]] = None) -> Any:
        """
        Close the handler and clean up any resources.

        Args:
            callback: Optional callback function to call when closing is complete

        Returns:
            An optional result that may be used by the driver
        """
        pass

    @abstractmethod
    def connect(self) -> Union[Awaitable[str], None]:
        """
        Connect to the southbound device.

        Returns:
            A string representing the connection status or None if using callbacks.
        """
        pass

    def cmd(self, command_name: Any, payload: Any) -> None:
        """
        Call a command on the southbound device.

        Args:
            command_name: Name of the command to execute. To be parsed by the handler.
            payload: Payload for the command to use. To be parsed by the handler.
        """
        pass

    @classmethod
    @abstractmethod
    def create(cls, driver, conf: Dict[str, Any]) -> 'Handler':
        """
        Create a new handler instance for the given driver and configuration.

        Args:
            driver: The driver instance that will use this handler
            conf: Configuration data for initializing the handler

        Returns:
            A new handler instance, or None if creation failed
        """
        pass

    def poll(self, spec: Any) -> Union[Any, Awaitable[Any]]:
        """
        Poll data from the southbound device for the given specification.

        This method is optional and only required for handlers used with PolledDriver.

        Args:
            spec: The parsed address specification to poll data from

        Returns:
            The polled data (can be bytes, string, or any serializable data),
            or an awaitable that resolves to the data. Returns None if no data available.
        """
        return None

File: lib/rs-edge-driver/src/handler.rs
use std::hash::Hash;

use async_trait::async_trait;
use bytes::Bytes;

use crate::driver::DriverHandle;
use crate::error::{ConnectError, HandlerError};

/// The trait that driver authors implement to define how to communicate
/// with a specific southbound device.
///
/// The library handles all MQTT communication with the edge agent;
/// implementors only need to deal with the device protocol itself.
///
/// # Driver modes
///
/// There are two modes of operation — choose one by overriding the
/// appropriate default method:
///
/// - **Polled**: override [`poll`](Handler::poll) to read data from
///   the device on demand when the edge agent requests it.
/// - **Async**: override [`subscribe`](Handler::subscribe) to set up
///   subscriptions, then push data back via
///   [`DriverHandle::publish`] when values arrive.
///
/// # Example
///
/// ```rust,ignore
/// use async_trait::async_trait;
/// use bytes::Bytes;
/// use rs_edge_driver::*;
///
/// struct ModbusHandler {
///     handle: DriverHandle,
///     host: String,
///     port: u16,
///     // ... client state
/// }
///
/// #[async_trait]
/// impl Handler for ModbusHandler {
///     type Addr = ModbusAddr;
///
///     fn create(
///         handle: DriverHandle,
///         config: serde_json::Value,
///     ) -> Result<Self, HandlerError> {
///         let host = config["host"].as_str()
///             .ok_or_else(|| HandlerError::new("missing 'host'"))?
///             .to_string();
///         let port = config["port"].as_u64()
///             .ok_or_else(|| HandlerError::new("missing 'port'"))? as u16;
///         Ok(Self { handle, host, port })
///     }
///
///     fn parse_addr(&self, raw: &str) -> Option<ModbusAddr> {
///         // parse "1,holding,100,10" into a typed address
///         todo!()
///     }
///
///     async fn connect(&mut self) -> Result<(), ConnectError> {
///         // establish TCP connection to the Modbus device
///         todo!()
///     }
///
///     async fn poll(&mut self, addr: &ModbusAddr) -> Option<Bytes> {
///         // read registers and return the raw bytes
///         todo!()
///     }
///
///     async fn close(&mut self) {
///         // disconnect from the device
///     }
/// }
/// ```
#[async_trait]
pub trait Handler: Send + Sized {
    /// A parsed, validated representation of a device address.
    ///
    /// The edge agent sends addresses as strings; [`parse_addr`](Handler::parse_addr)
    /// converts them into this type. Using an associated type gives
    /// compile-time safety — no downcasting needed.
    type Addr: Send + Sync + Hash + Eq + Clone;

    /// Construct a new handler from the configuration object sent
    /// by the edge agent.
    ///
    /// Return `Err` if the configuration is invalid — the driver will
    /// report status `CONF` to the edge agent.
    ///
    /// The [`DriverHandle`] can be stored and used later to push
    /// async data back to the driver via [`DriverHandle::publish`].
    fn create(handle: DriverHandle, config: serde_json::Value) -> Result<Self, HandlerError>;

    /// Parse and validate a raw address string from the edge agent.
    ///
    /// Return `None` if the address is not valid for this handler.
    /// Invalid addresses cause the driver to report status `ADDR`.
    fn parse_addr(&self, addr: &str) -> Option<Self::Addr>;

    /// Connect (or reconnect) to the southbound device.
    ///
    /// Return `Ok(())` on success. The driver will set status `UP`.
    /// Return `Err(ConnectError::ConnectionFailed)` or
    /// `Err(ConnectError::AuthFailed)` on failure — the driver will
    /// retry after the configured reconnect delay.
    async fn connect(&mut self) -> Result<(), ConnectError>;

    /// Clean up resources and disconnect from the device.
    ///
    /// Called before the handler is dropped — either due to
    /// reconfiguration or shutdown.
    async fn close(&mut self);

    /// **Polled mode**: read data for a single address from the device.
    ///
    /// Called when the edge agent sends a poll request. Return
    /// `Some(bytes)` with the raw data, or `None` if the read failed.
    ///
    /// The default implementation returns `None`. Override this for
    /// polled drivers.
    async fn poll(&mut self, _addr: &Self::Addr) -> Option<Bytes> {
        None
    }

    /// **Async mode**: set up subscriptions for the given addresses.
    ///
    /// Called after [`connect`](Handler::connect) succeeds and
    /// addresses have been configured. When data arrives, push it
    /// back to the driver via [`DriverHandle::publish`].
    ///
    /// Return `true` if all subscriptions were set up successfully.
    ///
    /// The default implementation returns `true` (no-op). Override
    /// this for async/event-driven drivers.
    async fn subscribe(&mut self, _addrs: &[Self::Addr]) -> bool {
        true
    }

    /// Handle a command sent from the edge agent.
    ///
    /// Commands arrive on `fpEdge1/{id}/cmd/{name}` and are
    /// forwarded here. Override this if the device supports commands.
    async fn cmd(&mut self, _command: &str, _payload: Bytes) {}
}
